modules/3d/include/opencv2/3d.hpp

```diff
@@ -468,77 +468,258 @@ can be found in:
+
+        // allocates jtj, jtb and other resources for objective function calculation, sets probeX to current X
+        virtual void prepareVars() = 0;
+        // returns a J^T*b vector (aka gradient)
+        virtual const Mat_<double> getJtb() = 0;
+        // returns a J^T*J diagonal vector
+        virtual const Mat_<double> getDiag() = 0;
+        // sets a J^T*J diagonal
+        virtual void setDiag(const Mat_<double>& d) = 0;
+        // performs jacobi scaling if the option is turned on
+        virtual void doJacobiScaling(const Mat_<double>& di) = 0;
+
+        // decomposes LevMarq matrix before solution
+        virtual bool decompose() = 0;
+        // solves LevMarq equation (J^T*J + lmdiag) * x = -right for current iteration using existing decomposition
+        // right can be equal to J^T*b for LevMarq equation or J^T*rvv for geodesic acceleration equation
+        virtual bool solveDecomposed(const Mat_<double>& right, Mat_<double>& x) = 0;
+
+        // calculates J^T*f(geo) where geo is geodesic acceleration variable
+        // this is used for J^T*rvv calculation for geodesic acceleration
+        // calculates J^T*rvv where rvv is second directional derivative of the function in direction v
+        // rvv = (f(x0 + v*h) - f(x0))/h - J*v)/h
+        // where v is a LevMarq equation solution
+        virtual bool calcJtbv(Mat_<double>& jtbv) = 0;
+
+        // sets current params vector to probe params
+        virtual void acceptProbe() = 0;
     };
 
+    // normalize jacobian columns for better conditioning
+    // slows down sparse solver, but maybe this'd be useful for some other solver
+    bool jacobiScaling;
+    // double upFactor until the probe is successful
+    bool upDouble;
+    // use stepQuality metrics for steps down
+    bool useStepQuality;
+    // clamp diagonal values added to J^T*J to pre-defined range of values
+    bool clampDiagonal;
+    // to use squared L2 norm or Inf norm for step size estimation
+    bool stepNormInf;
+    // to use relEnergyDeltaTolerance or not
+    bool checkRelEnergyChange;
+    // to use minGradientTolerance or not
+    bool checkMinGradient;
+    // to use stepNormTolerance or not
+    bool checkStepNorm;
+    // to use geodesic acceleration or not
+    bool geodesic;
+    // second directional derivative approximation step for geodesic acceleration
+    double hGeo;
+    // how much of geodesic acceleration is used
+    double geoScale;
+    // optimization stops when norm2(dx) drops below this value
+    double stepNormTolerance;
+    // optimization stops when relative energy change drops below this value
+    double relEnergyDeltaTolerance;
+    // optimization stops when max gradient value (J^T*b vector) drops below this value
+    double minGradientTolerance;
+    // optimization stops when energy drops below this value
+    double smallEnergyTolerance;
+    // optimization stops after a number of iterations performed
+    unsigned int maxIterations;
```

Comment:
No fields in public API.
This is the primary requirement of public OpenCV API.
getter/setter should be exposed instead.

Reply:
can we try to utilize https://github.com/opencv/opencv/wiki/OE-34.-Named-Parameters. It lets to define very nice API for C++ and Python (and maybe Swift after the corresponding binding generator is revised)

Reply:
Implemented it as a `Setting` structure